### DISCUSSION TITLE
A "next dev" option/mode, to compile all pages at once

### DISCUSSION THREAD

**Discussion by nicolaslima321** [vercel/next.js#51672] [Ideas]
### Goals

1. Have an option/flag to `next dev`, that compiles **all the pages**, inside the app directory, at once, instead of compiling on each request. This could be something like `next dev --compile-all-pages` for example.

### Non-Goals

1. Restrict the NextJs development, to compile the components only in a kind of route/request demand :confused:.

### Background

This feature is needed, because at least by me.. I'm having a very **bad developer experience**, when I need to navigate in many, or more than one pages in my application.
I understand the reasons, but, whats the matter on provide an option to compile all the pages? Some developers may prefer to wait until all the pages compile only once, instead of wait on the compile every single time.

### Proposal

Maybe implement a flag to `next dev` understand that it's supposed to compile all the pages

**Comment by nicolaslima321** [comment]
this feature request comes from the follow issue https://github.com/vercel/next.js/issues/6554

**Reply by FrenchMajesty** [reply]
I agree. This is much needed. I am new to the NextJS but the `next dev` is so unusable I am actually baffled that startups are building with this?!

It takes 10sec for me to transition from one page to the next.

**Reply by Divyam-Kumar-Pandey** [reply]
I have a hack for this-
But you'll have to wait a few more seconds after starting the server.
The hack is to create a .bat file and write the routes you are working on. (you can write all of them, but it will take more time)
let's say I'm working on the dashboard today.
like this:
curl http://localhost:3000/dashboard/home
curl http://localhost:3000/dashboard/help
curl http://localhost:3000/dashboard/something.....

Now just run your bat file just after starting your dev server.
Pro Tip: you can make files like dashboard-prerender.bat, landingpage.bat, api-routes.bat, all.bat for all different kind of needs.

**Comment by sgoodrow** [comment]
A hybrid option would be nice. Compile pages as fast as possible, starting with those requested by the client.

**Comment by jordanbtucker** [comment]
I just ran into a different use case for this feature.

I'm prototyping some REST calls using a mock `app/notes/db.ts` file. In that file, I have a `NOTES` array that mocks a DB collection and a few functions for reading and manipulating that array (`getNotes`, `createNote`, etc.).

Then, I have an `app/notes/route.ts` file with a `GET` and `POST` function, and an `app/notes/[id]/route.ts` file with a `GET` and `PUT` function, and they both call the mock DB functions. If I send a `GET` or `POST` request to `/notes`, the `NOTES` array is initialized on the fly, which is fine. But the first time I make a `GET` or `PUT` request to `/notes/1`, it compiles `app/notes/[id]/route.ts` and the array is reinitialized, so any notes I created before that are gone.

Obviously, this interferes with mocking my rest calls. Sure, I'll eventually be using tests, but I'm just prototyping right now, and it's frustrating that I have to make a request to both `/notes` and `/notes/1` before I can make any requests that don't reinitialize my mock DB. I'm used to my mock data being reinitialized after changing a file, but not in the middle of testing.

Maybe I'm missing something.

**Comment by CodeGeek04** [comment]
I agree, we need some kind of option in dev script to atleast specify the pages to be compiled beforehand, if not all of them.

**Comment by MathiasOxholm** [comment]
Agree with this one. I'm currently running Playwright with the dev server, and navigating between pages causes timeouts from Playwright's side. Bumping the timeout fixes it sort of, but would be nice not to have this issue at all.

**Comment by timmypidashev** [comment]
This would be very nice. Currently struggling with some pretty annoying issues like caddy giving a 502 because my next-auth page doesnt compile fast enough in dev mode. A feature flag to precompile in dev mode would be very neat and highly appreciated :+1:

**Comment by andrewmartin** [comment]
I'd love this as well.

I wonder if there's some weird or hacky way to do this using old fashioned `curl` and a combination of [`useSelectedLayoutSegments`.](https://nextjs.org/docs/app/api-reference/functions/use-selected-layout-segments) Idea would be to write a script that:

1. prints out all of the registered routes
2. then runs `curl -XGET` on all of them

Definitely hacky, but might get the job done?

**Comment by okunamayanad** [comment]
I'd love to see this. Currently, I just hit all the pages when I restart the dev server but this is not optimal for DX. I have a beefy CPU so I'd like an option to just compile all pages in the background

**Comment by ishchhabra** [comment]
For anyone facing the same issue, I have built this functionality in a fork at https://github.com/ishchhabra/next.js/tree/v14.2.18. The package is available precompiled for `v14.2.18` and can be used directly in `package.json` like 

`"next": "https://github.com/ishchhabra/next.js/releases/download/v14.2.18-precompile-routes/v14.2.18-precompile-routes.tgz"`

To use this, simply run the development server with `pnpm dev --precompile-routes`.

Would be happy to submit a PR if this is something that the Next.js team would be open to having as part of the project.

**Reply by Hova25** [reply]
Hello @ishchhabra i have tested it and it works very well
I think is necessary to add config to choose specific paths to precompile because otherwise it takes too much time

**Comment by daross-wizehire** [comment]
I'd also love this feature. Currently, trying to test animated page transitions in dev mode is difficult because of this issue. Navigating to a new route via `router.push` should update the `{children}`, but with JIT compilation, the transition happens while the route is still compiling. The new page then re-renders partway through the animation or after it's done, causing it to appear broken.

**Comment by Hova25** [comment]
To wait, you can use a little hack..

in file node_modules/next/dist/server/dev/next-dev-server.js

you can add after  `await super.prepareImpl();`

```
await this.matchers.reload();
for (const matcher of [
    ...this.matchers.matchers.static,
    ...this.matchers.matchers.dynamic,
].filter( m => YOUR FILTER FUNCTION based  m.definition.pathname pathname is routing from app directory start with / ) ) {
    try {
        await this.ensureMiddleware(matcher.definition.page)
        await this.ensurePage({
            definition: matcher.definition,
            page: matcher.definition.page,
            clientOnly: false,
            url: matcher.definition.page,
        })
    } catch (err) {
        console.error(`Failed to precompile route ${matcher.definition.page}:`)
    }
}
```

If you want you can not implement the filter function and it compile all your app but it takes a long time

Thank you, @ishchhabra, for guiding me with your PR

**Comment by Walidadebayo** [comment]
I solved this problem by using puppeteer to go through all my route after running `npm run dev` using concurrently package to both command at the same time

![Image](https://github.com/user-attachments/assets/a9c11c02-b732-49f0-85fd-4f8b55d7c85c)

// precompile.js
import puppeteer from 'puppeteer';
import http from 'http';

const routes = [
//routes
    '/',
];


const checkServer = (url) => {

    return new Promise((resolve, reject) => {
        const req = http.get(url, (res) => {
            if (res.statusCode === 200) {
                resolve(true);
            } else {
                reject(false);
            }
        });

        req.on('error', () => resolve(false));
        req.end();
    });
};

(async () => {

    const serverUrl = 'http://localhost:3500';
    const serverRunning = await checkServer(serverUrl);
    if (!serverRunning) {
        console.error('Server is not running. Please start the server and try again.');
        process.exit(1);
    }

    const browser = await puppeteer.launch();
    const page = await browser.newPage();
    page.setDefaultNavigationTimeout(15000);
    page.setDefaultTimeout(15000);

    const retryVisit = async (url, retries = 3) => {
        for (let attempt = 1; attempt <= retries; attempt++) {
            try {
                await page.goto(url, { waitUntil: 'networkidle0' });
                return;
            } catch (error) {
                if (attempt === retries) {
                    console.error(`Failed to visit ${url} after ${retries} attempts: ${error.message}`);
                } else {
                    console.log(`Retrying ${url} (attempt ${attempt} of ${retries})`);
                }
            }
        }
    };

    for (const route of routes) {
        const url = `${serverUrl}${route}`;
        console.log(`Visiting ${url}`);
        await retryVisit(url);
    }

    await browser.close();
})();

**Reply by felipires** [reply]
Can't you just send a GET request for each route instead of wrapping everything in a whole browser? Take a look at this [approach](https://github.com/vercel/next.js/discussions/51672#discussioncomment-12541156)

**Comment by Divyam-Kumar-Pandey** [comment]
I have a hack for this-
But you'll have to wait a few more seconds after starting the server.
The hack is to create a .bat file and write the routes you are working on. (you can write all of them, but it will take more time)
let's say I'm working on the dashboard today.
like this:
curl http://localhost:3000/dashboard/home
curl http://localhost:3000/dashboard/help
curl http://localhost:3000/dashboard/something.....

Now just run your bat file just after starting your dev server.
Pro Tip: you can make files like dashboard-prerender.bat, landingpage.bat, api-routes.bat, all.bat for all different kind of needs.

**Comment by machineghost** [comment]
One more use case: end-to-end tests.  It's a hassle to have to build and run your app every time you want to run your E2E tests ... but because you can't predict when the dev server will be ready, it's effectively unusable for E2E testing.

If there was a way to skip the build/run step, but still have a somewhat predictable server, it would make such testing a lot more convenient.

**Comment by debuglevel** [comment]
Would like to see this.

Launching `next dev` and see a `✓ Ready in 2.5s` seems like a bad joke when a `✓ Compiled / in 14.3s` is following *after* I requested a page. That thing had plenty of time to do anything. It does not have to idle around after starting. Especially with something as slow and bulky as TypeScript/JavaScript development where thousand of files have to be compiled every time even when there was not a single change.

**Comment by oskarscholander** [comment]
Please add this!

**Comment by zulfio** [comment]
We need this feature

**Comment by marcoscannabrava** [comment]
3 years later we still need this. Badly. Even on Mac M4 or Ryzen AI Max 390 (tested on both).

Curious what people's workarounds are? Especially with AI agents.

**Reply by marcoscannabrava** [reply]
Honestly I changed my mind about this. I have found that large projects tend to end up with huge cache folders. My workaround is to run everything in containers*, destroy them from time to time, and just run `next build/start`.

> *there's no performance penalty on Linux and on Mac/Windows you can just delete the cache folder

**Comment by safareli** [comment]
lazy compilation is sometimes nice but sometimes on does need dev server in a ready state  to have better feel of more realistic end perf

**Reply by debuglevel** [reply]
Or to inhibit personal insanity. 🫠

**Comment by cni13** [comment]
I would love if there are two modes: wait for full pre compile until serving or pre compile after starting serving. Then people can decide which mode they like for their scenario.
And then an optional feature to include only routes you specify and then they are precompiled in order. 
Then imo you have most grounds covered concerncing this issue.
